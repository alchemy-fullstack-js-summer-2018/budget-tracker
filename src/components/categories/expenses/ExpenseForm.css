:local(.expenseForm) {
  border-radius: 10px;
  border: 3px solid red;
  width: 77%;
  margin-left: 10%;
  margin-right: 10%;
  text-align: center; 
  text-transform: capitalize;
}


