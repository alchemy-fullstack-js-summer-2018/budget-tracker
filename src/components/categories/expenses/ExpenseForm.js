import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ExpenseForm extends Component {
  state = {
    name: '',
    price: ''
  };

  static propTypes = {
    expense: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { expense } = this.props;
    if(!expense) return;
    this.setState(expense);
  }

  handleSubmit = event => {
    event.preventDefault();
    
    this.props.onComplete(this.state);
    this.setState({ name: '', price: '' });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { name, price } = this.state;
    const { onCancel, expense } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="name" value={name} onChange={this.handleChange}/>
        <InputControl name="price" value={price} type="number" onChange={this.handleChange}/>
        <p>
          <button type="submit">{ expense ? 'Update' : 'Add' }</button>
          {expense && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = props => (
  <p>
    <label>
      {props.name}:
      <input {...props} required />
    </label>
  </p>
);
 
export default ExpenseForm;