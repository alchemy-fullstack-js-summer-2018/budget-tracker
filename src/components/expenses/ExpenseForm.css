:local(.expenseForm) {
  font-size: 14px;
  padding: 3px;

  input {
    margin: 2px;
  }

  p {
    position: relative;
    left: 5%;
  }

  button {
    position: relative;
    left: 2%
  }
}