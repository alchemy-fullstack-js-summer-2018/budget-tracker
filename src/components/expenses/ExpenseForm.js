import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ExpenseForm extends Component {

  state = {
    key: null,
    categoryId: '',
    name: '',
    price: '',
    timestamp: null
  };

  static propTypes = {
    expense: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func,
    categoryId: PropTypes.string,
    
  };

  componentDidMount() {
    const { expense } = this.props;
    if(!expense) return;
    this.setState(expense);
  }

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  handleSubmit = (e) => {
    e.preventDefault();

    const { categoryId } = this.props;
    // this.setState({ categoryKey: categoryId });
    const { key, name, price, timestamp } = this.state;
    const expense = { key, name, categoryId, price, timestamp };
    
    if(key) expense.key = key;
    if(!key) expense.categoryId = categoryId;
    
    this.props.onComplete(expense);
    this.setState({ name: '', price: '' });
  };
  
  render() { 
    const { key, name, price } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <label>Expense name:<input type="text" name="name" value={name} onChange={this.handleChange}></input></label>
        <label>Price:<input type="number" name="price" value={price} onChange={this.handleChange}></input></label>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}
 
export default ExpenseForm;
