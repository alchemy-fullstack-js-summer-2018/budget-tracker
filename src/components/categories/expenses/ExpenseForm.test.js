import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseForm from './ExpenseForm';

describe('Expense Form component', () => {
  it('renders add if no expense prop', () => {
    const handleComplete = jest.fn();
    const wrapper = mount(<ExpenseForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const expense = { name: 'bus', price: 5 };

    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: expense.name
      }
    });

    wrapper.find('input[name="price"]').simulate('change', {
      target: {
        name: 'price',
        value: expense.price
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(expense);
  });

  it('renders edit if there is expense prop', () => {
    const handleComplete = jest.fn();
    const handleCancel = jest.fn();

    const expense = { name: 'bus', price: 5, categoryId: 'abc' };

    const wrapper = mount(<ExpenseForm
      onComplete={handleComplete}
      onCancel={handleCancel}
      expense={expense}
    />);

    expect(toJSON(wrapper)).toMatchSnapshot();

    wrapper.find('input[name="price"]').simulate('change', {
      target: {
        name: 'price',
        value: 3
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual({
      ...expense,
      price: 3
    });
  });
});