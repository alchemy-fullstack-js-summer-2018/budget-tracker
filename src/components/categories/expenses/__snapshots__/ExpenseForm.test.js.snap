// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense Form component renders add if no expense prop 1`] = `
<ExpenseForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="price"
      onChange={[Function]}
      type="number"
      value=""
    >
      <p>
        <label>
          price
          :
          <input
            name="price"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</ExpenseForm>
`;

exports[`Expense Form component renders edit if there is expense prop 1`] = `
<ExpenseForm
  expense={
    Object {
      "categoryId": "abc",
      "name": "bus",
      "price": 5,
    }
  }
  onCancel={[MockFunction]}
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="bus"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="bus"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="price"
      onChange={[Function]}
      type="number"
      value={5}
    >
      <p>
        <label>
          price
          :
          <input
            name="price"
            onChange={[Function]}
            required={true}
            type="number"
            value={5}
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Cancel
      </button>
    </p>
  </form>
</ExpenseForm>
`;
